doc/design/logging.md

```diff
@@ -0,0 +1,396 @@
+
+7. The default logger shall be replaceable via the platform abstraction in order
+   to use the dedicated logger for log messages which are emitted before it could
+   be replaced at runtime, e.g. from global objects before `main`.
+
+8. Additionally, there shall be a compile time option to forward all log messages
+   independent of the log level threshold for cases where the logger framework
+   does the filtering itself.
+
+9. For tests the default logger shall be replaced to suppress the log messages in
+   passed tests but output them for failed tests. This keeps the noise low but
+   still provides useful information when needed. Additionally there shall be an
+   environment variable to circumvent the suppression when desired.
+
+### Solution
+
+#### Class diagram
+
+![logging class diagram](../website/images/logging_classes.svg)
+
+The logger can be customized at compile time and at runtime. The former is done
+by the `Impl` template parameter and the latter by deriving from the logger.
+
+#### Logging with LogStream
+
+![logging with logstream](../website/images/logging_with_logstream.svg)
+
+#### Macro with lazy evaluation
+
+The buildup of the log message is only done when the log level is above the output
+threshold. This is accomplished by a macro with an incomplete if-statement.
+
+```cpp
+#define LAZY() if (cond)
+
+LAZY() expensiveFunctionCall();
+```
+
+In the example above `expensiveFunctionCall` is only executed if `cond` is `true`.
+If `cond` is a compile time constant set to `false`, the whole statement is compiled
+to a no-op and optimized away.
+
+This is the log macro with lazy evaluation
+
+```cpp
+#define IOX_LOG_INTERNAL(file, line, function, level)          \
+    if ((level) <= iox::log::Logger::minimalLogLevel()         \
+        && (iox::log::Logger::ignoreLogLevel()                 \
+            || (level) <= iox::log::Logger::getLogLevel()))    \
+    iox::log::LogStream(file, line, function, level).self()
+
+#define IOX_LOG(level) IOX_LOG_INTERNAL(__FILE__, __LINE__, __FUNCTION__, iox::log::LogLevel::level)
+```
+
+With `minimalLogLevel` and `ignoreLogLevel` being static constexpr functions
+the compiler will optimize this either to `if (false) iox::log::LogStream(...)`
+and finally completely away or
+`if ((level) <= iox::log::Logger::getLogLevel()) iox::log::LogStream(...)`.
+The minimal log level check is intended to fully optimize away a log statement
+and the ignore active log level check to always forward the log message to the
+logger, independent of the active log level.
+
+The `self()` invocation is used to create an lvalue reference to the `LogStream`
+object. This eases the implementation of logging support for custom types since
+`IOX_LOG(INFO) << myType;` would require to implement an overload with a rvalue
+`LogStream` reference but `IOX_LOG(INFO) << "#### " << myType;` requires a
+lvalue reference.
+
+The `IOX_LOG` macro is intended for general use and the `IOX_LOG_INTERNAL` for
```

Comment:
Here it would absolutely make sense to add a struct `SourceCodeOrigin` in the logger which bundles line, file, function together.
This struct then could be used in `Expects` and `Ensures` and would bring some uniformity.

Reply:
Already answered in https://github.com/eclipse-iceoryx/iceoryx/pull/1358#discussion_r906194754

It is an implementation detail and can easily be done once we know where we can put such a struct. It is not necessary to have it in the initial implementation.